tests/zfs-tests/cmd/crypto_test.c

```diff
@@ -0,0 +1,1246 @@
+		foreach_aes_ccm(run_test_alg_cb, &args, outmode);
+		break;
+	case ALG_AES_GCM:
+		foreach_aes_gcm(run_test_alg_cb, &args, outmode);
+		break;
+	default:
+		abort();
+	}
+
+	icp_fini();
+
+	return (args.failed);
+}
+
+/* ========== */
+
+/* performance tests */
+
+/* helper; fill the given buffer with random data */
+static int
+fill_random(uint8_t *v, size_t sz)
+{
+	int fd = open("/dev/urandom", O_RDONLY);
+	if (fd < 0)
+		return (errno);
+
+	while (sz > 0) {
+		ssize_t r = read(fd, v, sz);
+		if (r < 0) {
+			close(fd);
+			return (errno);
+		}
+		v += r;
+		sz -= r;
+	}
+
+	close(fd);
+
+	return (0);
+}
+
+/* args for perf_alg_cb */
+typedef struct {
+	crypto_test_alg_t	alg;
+	uint8_t			*msg;
+	uint8_t			*out;
+	uint8_t			key[32];
+	uint8_t			iv[12];
+} perf_alg_args_t;
+
+#define	PERF_MSG_SHIFT_MIN	(10)	/* min test size 2^10 == 1K */
+#define	PERF_MSG_SHIFT_MAX	(19)	/* max test size 2^19 == 512K */
+#define	PERF_ROUNDS		(1000)	/* 1000 rounds per test */
+
+/* per-alg-impl function for performance test runs */
+static void
+perf_alg_cb(const char *alginfo, void *arg)
+{
+	char buf[10];
+	printf("%-28s", alginfo);
+
+	if (arg == NULL) {
+		printf("[not supported on this platform]\n");
+		return;
+	}
+
+	perf_alg_args_t *args = arg;
+
+	/* space for mechanism description */
+	crypto_mechanism_t mech = {};
+	union {
+		CK_AES_GCM_PARAMS gcm;
+		CK_AES_CCM_PARAMS ccm;
+	} params = {};
+	mech.cm_param = (caddr_t)&params;
+
+	/* loop for each power-2 input size */
+	for (int i = PERF_MSG_SHIFT_MIN; i <= PERF_MSG_SHIFT_MAX; i++) {
+		/* size of input */
+		size_t sz = 1<<i;
+
+		/* initialise mechanism */
+		init_mech(&mech, args->alg, args->iv, sizeof (args->iv),
+		    val_empty, 0, sz, 16, B_FALSE);
+
+		/* run N rounds and accumulate total time */
+		uint64_t total = 0;
+		for (int round = 0; round < PERF_ROUNDS; round++) {
+			uint64_t usec;
+			encrypt_one(&mech, args->key, sizeof (args->key),
+			    args->msg, sz, args->out, sz+16, &usec);
+			total += usec;
+		}
+
+		/*
+		 * print avg time per round. zfs_nicetime expects nanoseconds,
+		 * so we multiply first
+		 */
+		zfs_nicetime((total*1000)/PERF_ROUNDS, buf, sizeof (buf));
+		printf("  %5s", buf);
+	}
+
+	printf("\n");
+}
+
+/* main function for performance tests */
+static int
+perf_main(const char *algname, crypto_test_outmode_t outmode)
+{
+	perf_alg_args_t args;
+
+	if (strcmp(algname, "AES-CCM") == 0)
+		args.alg = ALG_AES_CCM;
+	else if (strcmp(algname, "AES-GCM") == 0)
+		args.alg = ALG_AES_GCM;
+	else {
+		fprintf(stderr, "E: unknown algorithm: %s\n", algname);
+		return (1);
+	}
+
+	/*
+	 * test runs are often slow, but the very first ones won't be. by
+	 * disabling buffering, we can display results immediately, and
+	 * the user quickly gets an idea of what to expect
+	 */
+	setvbuf(stdout, NULL, _IONBF, 0);
+
+	/* allocate random data for encrypt input */
+	size_t maxsz = (1<<PERF_MSG_SHIFT_MAX);
+	args.msg = malloc(maxsz);
+	fill_random(args.msg, maxsz);
```

Comment:
nit:  add check/assert that `fill_random(...)` returns 0

Reply:
Fair call!